src/util/LockUtils.ts

```diff
@@ -0,0 +1,102 @@
+import { getLoggerFor } from '../logging/LogUtil';
+import { InternalServerError } from './errors/InternalServerError';
+
+const logger = getLoggerFor('LockUtil');
+
+/**
+ * Await this function to wait a set amount of time (without consuming cpu).
+ * @param delay - How long to wait.
+ * @param jitter - A fraction of this jitter will be added to the delay.
+ * @returns Promise<void> when ready.
+ */
+export async function setJitterTimeout(delay: number, jitter = 0): Promise<void> {
+  function calcTime(): number {
+    jitter = jitter > 0 ? Math.max(1, Math.floor(Math.random() * jitter)) : 0;
+    return Math.max(0, delay + jitter);
+  }
+  return new Promise<void>((resolve): any => setTimeout(resolve, calcTime()));
+}
+
+export interface AttemptSettings {
+  /** How many times should an operation be retried. (-1 is indefinitely). */
+  retryCount?: number;
+  /** The how long should the next retry be delayed (+ some retryJitter) (in ms). */
+  retryDelay?: number;
+  /** Add a fraction of jitter to the original delay each attempt (in ms). */
+  retryJitter?: number;
+}
+
+export interface LockRequestState<T> {
+  /** Should the function be retried. This means retrying might change the outcome. */
+  shouldRetry: boolean;
+  /** If shouldRetry is true, than a response might me present */
+  response?: T;
+  /** An error might be present, no matter if shouldRetry is true or false. */
+  error?: Error;
+}
+
+/**
+ * Will execute the given function until one of the following 3 cases occurs:
+ * * The function resolves: the resolved Promise is returned.
+ * * The function errors: the rejected error is returned.
+ * * The function did not resolve after the set amount of retries:
+ *   the rejected error is returned.
+ * @param fn - The function to retry. You can wrap the function with {@link toRetryOnBoolean}
+ *              or {@link toRetryOnError} to customize when-to-retry behaviour
+ * @param settings - The options on how to retry the function
+ */
+export async function retryFunctionUntil<T>(fn: () => Promise<LockRequestState<T>>,
```

Comment:
Took me some time to fully get how everything worked together 😅 . But now that I do, I'm not sure the request state has much added value. If the state contains error, it will get re-thrown immediately, and if it has a result, it will be returned immediately. Unless `shouldRetry` is set to true, but having `shouldRetry` be true and also one of the other fields filled feels like `fn` has been badly constructed. What if instead of using a request state, instead there is something like
```ts
let result = await fn();
while (typeof result === 'undefined' && tries < maxTries) {
 await setJitterTimeout(retryDelay, retryJitter);
 result = await fn();
 tries += 1;
}
// Return result if defined, otherwise throw max tries error
```
That way, if `fn` throws an error, that will still be thrown, and the result will still be returned. This way it's easier to provide existing functions to `retryFunctionUntil` without always needing a wrapper should it not be necessary. This does mean `retryFunctionUntil` only works with functions that return a value, but that is also easily solvable with a wrapper function like those below should it really be necessary.